pkg/actors/internal/placement.go

```diff
@@ -0,0 +1,374 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package internal
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	dapr_credentials "github.com/dapr/dapr/pkg/credentials"
+	diag "github.com/dapr/dapr/pkg/diagnostics"
+	"github.com/dapr/dapr/pkg/logger"
+	"github.com/dapr/dapr/pkg/placement/hashing"
+	v1pb "github.com/dapr/dapr/pkg/proto/placement/v1"
+	"github.com/dapr/dapr/pkg/runtime/security"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var log = logger.NewLogger("dapr.runtime.actor.internal.placement")
+
+const (
+	lockOperation   = "lock"
+	unlockOperation = "unlock"
+	updateOperation = "update"
+
+	placementReconnectInterval    = 100 * time.Millisecond
+	statusReportHeartbeatInterval = 1 * time.Second
+)
+
+// ActorPlacement maintains membership of actor instances and consistent hash
+// tables to discover the actor while interacting with Placement service.
+type ActorPlacement struct {
+	actorTypes []string
+	appID      string
+	// runtimeHostname is the address and port of the runtime
+	runtimeHostName string
+
+	// serverAddr is the list of placement addresses.
+	serverAddr []string
+	// serverIndex is the the current index of placement servers in serverAddr.
+	serverIndex int
+	// streamConnAlive is the status of stream connection alive.
+	streamConnAlive bool
+	// streamConnectedCh is the channel to notify that the stream
+	// between runtime and placement is connected.
+	streamConnectedCh chan bool
+	// clientCert is the workload certificate to connect placement.
+	clientCert *dapr_credentials.CertChain
+	// clientConn is the gRPC client connection.
+	clientConn *grpc.ClientConn
+	// clientStream is the client side stream.
+	clientStream v1pb.Placement_ReportDaprStatusClient
+
+	// placementTables is the consistent hashing table map to
+	// look up Dapr runtime host addressto locate actor.
+	placementTables *hashing.ConsistentHashTables
+	// placementTableLock is the lock for placementTables.
+	placementTableLock *sync.RWMutex
+
+	// unblockSignal is the channel to unblock table locking.
+	unblockSignal chan struct{}
+	// tableIsBlocked is the status of table lock.
+	tableIsBlocked bool
+	// operationUpdateLock is the lock for three stage commit.
+	operationUpdateLock *sync.Mutex
+
+	// appHealthFn is the user app health check callback.
+	appHealthFn func() bool
+	// afterTableUpdateFn is function for post processing done after table updates,
+	// such as draining actors and resetting reminders.
+	afterTableUpdateFn func()
+
+	// shutdown is the flag when runtime is being shutdown.
+	shutdown bool
+	// shutdownConnLoop is the wait group to wait until all connection loop are done
+	shutdownConnLoop sync.WaitGroup
+}
+
+func addDNSResolverPrefix(addr []string) []string {
+	resolvers := []string{}
```

Comment:
```suggestion
	resolvers := make([]string, len(resolvers))
```
(preallocate since we know the size)